src/wp-includes/blocks.php

```diff
@@ -1596,10 +1596,26 @@ function make_after_block_visitor( $hooked_blocks, $context, $callback = 'insert
  *
  * @since 5.3.1
  *
- * @param array $block_attributes Attributes object.
+ * @param array       $block_attributes Attributes object.
+ * @param string|null $block_name       Block name. Null if the block name is unknown. Optional.
  * @return string Serialized attributes.
  */
-function serialize_block_attributes( $block_attributes ) {
+function serialize_block_attributes( $block_attributes, $block_name = null ) {
+	foreach ( $block_attributes as $attribute => $value ) {
+		if ( is_array( $value ) && empty( $value ) ) {
+			// An empty `array()` is encoded as `[]` in JSON. However, it's possible
+			// that the attribute type is really an object (associative array in PHP),
+			// so we need to check for that.
+			$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block_name );
```

Comment:
Nit: It's probably better to move this outside of the loop. There's no reason to get a block type instance on each iteration.

Reply:
Ohh that's right 😅 
The reason I put it there was to avoid the block type lookup altogether if there was no empty array at all. But we can probably just cache it after it's first been looked up.

Reply:
The registry is already working like a memory cache, right? The lookup shouldn't be a problem.

Reply:
81515324bc5a53a25bb4adba4d6df8eb5f1376ee